File: research-assistant/backend/app/schemas/retrieval.py

from __future__ import annotations

from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ---------------------------------------------------------------------------
# Query Analysis
# ---------------------------------------------------------------------------


class QueryAnalysis(BaseModel):
    """
    Structured analysis of the user's retrieval query.
    """

    model_config = ConfigDict(from_attributes=True)

    original_query: str
    normalized_query: str | None = None

    keywords: list[str] = Field(default_factory=list)
    entities: list[str] = Field(default_factory=list)
    concepts: list[str] = Field(default_factory=list)

    filters: dict[str, Any] = Field(default_factory=dict)

    requires_semantic_search: bool = True
    requires_keyword_search: bool = True


class QueryClassification(BaseModel):
    """
    Classification of a retrieval query.
    """

    model_config = ConfigDict(from_attributes=True)

    query_type: str = "general"
    intent: str = "information_retrieval"

    domain: str | None = None

    confidence: float = Field(
        default=0.0,
        ge=0.0,
        le=1.0,
    )

    labels: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Provenance
# ---------------------------------------------------------------------------


class Provenance(BaseModel):
    """
    Source information for a retrieved chunk.

    Important ID conventions:

    - papers.id -> INTEGER
    - document_chunks.id -> UUID
    - document_chunks.document_id -> UUID
    """

    model_config = ConfigDict(from_attributes=True)

    # papers.id is INTEGER in the database.
    paper_id: int | None = None

    # document/document-chunk identifiers are UUIDs.
    document_id: UUID | None = None

    chunk_id: UUID

    page: int | None = None
    section: str | None = None
    source: str | None = None


# ---------------------------------------------------------------------------
# Evidence
# ---------------------------------------------------------------------------


class Evidence(BaseModel):
    """
    Evidence associated with a retrieved result.

    This provides the retrieval layer with a stable structure that can
    later be consumed by citation, reasoning, and report-generation layers.
    """

    model_config = ConfigDict(from_attributes=True)

    evidence_id: UUID | None = None

    content: str

    score: float = 0.0

    rank: int | None = None

    provenance: Provenance

    metadata: dict[str, Any] = Field(
        default_factory=dict
    )


# ---------------------------------------------------------------------------
# Retrieval Request
# ---------------------------------------------------------------------------


class RetrievalRequest(BaseModel):
    """
    Request payload for knowledge retrieval.

    ID conventions:

    - paper_id -> papers.id -> INTEGER
    - document_id -> UUID
    """

    model_config = ConfigDict(from_attributes=True)

    query: str = Field(
        ...,
        min_length=1,
        max_length=5000,
        description="Natural-language research query.",
    )

    # papers.id is INTEGER.
    paper_id: int | None = Field(
        default=None,
        description="Restrict retrieval to a specific paper ID.",
    )

    # document_id is UUID.
    document_id: UUID | None = Field(
        default=None,
        description="Restrict retrieval to a specific document.",
    )

    top_k: int = Field(
        default=10,
        ge=1,
        le=100,
        description="Maximum number of results.",
    )

    vector_weight: float = Field(
        default=0.7,
        ge=0.0,
        le=1.0,
        description="Weight assigned to vector similarity.",
    )

    keyword_weight: float = Field(
        default=0.3,
        ge=0.0,
        le=1.0,
        description="Weight assigned to keyword relevance.",
    )


# ---------------------------------------------------------------------------
# Retrieval Result
# ---------------------------------------------------------------------------


class RetrievalResult(BaseModel):
    """
    Canonical representation of one retrieved knowledge chunk.

    This is the primary backend/frontend retrieval contract.

    ID conventions:

    - chunk_id -> UUID
    - document_id -> UUID
    - paper_id -> INTEGER
    - section_id -> UUID
    """

    model_config = ConfigDict(from_attributes=True)

    # -----------------------------------------------------------------------
    # Identity
    # -----------------------------------------------------------------------

    chunk_id: UUID

    document_id: UUID

    # papers.id is INTEGER, not UUID.
    paper_id: int | None = None

    section_id: UUID | None = None

    # -----------------------------------------------------------------------
    # Content
    # -----------------------------------------------------------------------

    content: str

    # -----------------------------------------------------------------------
    # Ranking
    # -----------------------------------------------------------------------

    score: float = 0.0

    rank: int = 0

    retrieval_method: str = "hybrid"

    rerank_score: float | None = None

    # -----------------------------------------------------------------------
    # Individual retrieval scores
    # -----------------------------------------------------------------------

    vector_score: float | None = None

    keyword_score: float | None = None

    # -----------------------------------------------------------------------
    # Source information
    # -----------------------------------------------------------------------

    provenance: Provenance | None = None

    # -----------------------------------------------------------------------
    # Optional evidence representation
    # -----------------------------------------------------------------------

    evidence: Evidence | None = None

    # -----------------------------------------------------------------------
    # Extensible metadata
    # -----------------------------------------------------------------------

    metadata: dict[str, Any] = Field(
        default_factory=dict
    )


# ---------------------------------------------------------------------------
# Retrieval Response
# ---------------------------------------------------------------------------


class RetrievalResponse(BaseModel):
    """
    Canonical retrieval API response.
    """

    model_config = ConfigDict(from_attributes=True)

    query: str

    results: list[RetrievalResult] = Field(
        default_factory=list
    )

    total: int = 0

    top_k: int = 10

    retrieval_mode: str = "hybrid"

    # -----------------------------------------------------------------------
    # Query understanding
    # -----------------------------------------------------------------------

    query_analysis: QueryAnalysis | None = None

    query_classification: QueryClassification | None = None

    # -----------------------------------------------------------------------
    # Optional response-level evidence
    # -----------------------------------------------------------------------

    evidence: list[Evidence] = Field(
        default_factory=list
    )

    # -----------------------------------------------------------------------
    # Optional diagnostics
    # -----------------------------------------------------------------------

    metadata: dict[str, Any] = Field(
        default_factory=dict
    )

File: research-assistant/frontend/src/types/retrieval.ts
/**
 * Canonical frontend types for the unified retrieval system.
 *
 * Retrieval supports:
 *
 *   semantic
 *   keyword
 *   graph
 *   hybrid
 *   fusion
 *   reranking
 *   evidence
 *   provenance
 *
 * Graph retrieval is a capability of the retrieval engine.
 *
 * It is NOT a separate Graph RAG API.
 */

/* -------------------------------------------------------------------------- */
/* Shared                                                                      */
/* -------------------------------------------------------------------------- */

export type UUID = string;

export type RetrievalMetadata = Record<string, unknown>;

export type RetrievalMethod =
  | "semantic"
  | "keyword"
  | "graph"
  | "hybrid"
  | "fusion"
  | "unknown"
  | string;

export type GraphNodeType =
  | "paper"
  | "author"
  | "method"
  | "dataset"
  | "topic"
  | "organization"
  | "venue"
  | "document"
  | "section"
  | "chunk"
  | "concept"
  | string;

/* -------------------------------------------------------------------------- */
/* Query Analysis                                                              */
/* -------------------------------------------------------------------------- */

export interface QueryAnalysis {
  original_query: string;

  normalized_query: string | null;

  keywords: string[];

  entities: string[];

  concepts: string[];

  filters: RetrievalMetadata;

  requires_semantic_search: boolean;

  requires_keyword_search: boolean;

  requires_graph_search?: boolean;

  query_type?: string | null;

  intent?: string | null;

  domain?: string | null;

  complexity?: number | null;
}

/* -------------------------------------------------------------------------- */
/* Classification                                                              */
/* -------------------------------------------------------------------------- */

export interface QueryClassification {
  query_type: string;

  intent: string;

  domain: string | null;

  confidence: number;

  labels: string[];
}

/* -------------------------------------------------------------------------- */
/* Graph provenance                                                            */
/* -------------------------------------------------------------------------- */

export interface GraphPathNode {
  id: UUID;

  type: GraphNodeType;

  label?: string | null;

  metadata?: RetrievalMetadata;
}

export interface GraphRelationship {
  id?: UUID | null;

  type: string;

  source_node_id: UUID;

  target_node_id: UUID;

  score?: number | null;

  metadata?: RetrievalMetadata;
}

/* -------------------------------------------------------------------------- */
/* Provenance                                                                  */
/* -------------------------------------------------------------------------- */

export interface Provenance {
  paper_id: UUID | null;

  document_id: UUID | null;

  chunk_id: UUID;

  page: number | null;

  section: string | null;

  section_id?: UUID | null;

  source: string | null;

  graph_node_id?: UUID | null;

  graph_node_type?: GraphNodeType | null;

  graph_edge_id?: UUID | null;

  graph_edge_type?: string | null;

  graph_path?: GraphPathNode[] | null;
}

/* -------------------------------------------------------------------------- */
/* Graph Evidence                                                              */
/* -------------------------------------------------------------------------- */

export interface GraphEvidence {
  evidence_id: UUID | null;

  content: string;

  score: number;

  rank: number | null;

  node_id: UUID | null;

  node_type: GraphNodeType | null;

  node_label: string | null;

  edge_id: UUID | null;

  edge_type: string | null;

  path: GraphPathNode[];

  relationships?: GraphRelationship[];

  provenance: Provenance | null;

  metadata: RetrievalMetadata;
}

/* -------------------------------------------------------------------------- */
/* Unified Evidence                                                            */
/* -------------------------------------------------------------------------- */

export interface Evidence {
  evidence_id: UUID | null;

  content: string;

  score: number;

  rank: number | null;

  provenance: Provenance;

  retrieval_method?: RetrievalMethod;

  graph?: GraphEvidence | null;

  metadata: RetrievalMetadata;
}

/* -------------------------------------------------------------------------- */
/* Retrieval Request                                                           */
/* -------------------------------------------------------------------------- */

export interface RetrievalRequest {
  query: string;

  paper_id?: UUID | null;

  document_id?: UUID | null;

  top_k?: number;

  vector_weight?: number;

  keyword_weight?: number;

  graph_enabled?: boolean;

  graph_depth?: number;

  graph_top_k?: number;

  retrieval_methods?: RetrievalMethod[];

  filters?: RetrievalMetadata;
}

/* -------------------------------------------------------------------------- */
/* Retrieval Result                                                            */
/* -------------------------------------------------------------------------- */

export interface RetrievalResult {
  chunk_id: UUID;

  document_id: UUID;

  paper_id: UUID | null;

  section_id: UUID | null;

  content: string;

  score: number;

  rank: number;

  retrieval_method: RetrievalMethod;

  rerank_score: number | null;

  vector_score: number | null;

  keyword_score: number | null;

  graph_score?: number | null;

  provenance: Provenance | null;

  evidence: Evidence | null;

  graph?: GraphEvidence | null;

  metadata: RetrievalMetadata;
}

/* -------------------------------------------------------------------------- */
/* Capability-specific results                                                */
/* -------------------------------------------------------------------------- */

export type SemanticRetrievalResult = RetrievalResult;

export type KeywordRetrievalResult = RetrievalResult;

export interface GraphRetrievalResult extends RetrievalResult {
  retrieval_method: "graph" | string;

  graph: GraphEvidence | null;

  graph_score: number | null;
}

/* -------------------------------------------------------------------------- */
/* Retrieval Response                                                          */
/* -------------------------------------------------------------------------- */

export interface RetrievalResponse {
  query: string;

  /**
   * Final fused/reranked results.
   */
  results: RetrievalResult[];

  semantic_results: SemanticRetrievalResult[];

  keyword_results: KeywordRetrievalResult[];

  graph_results: GraphRetrievalResult[];

  total: number;

  top_k: number;

  retrieval_mode: string;

  retrieval_methods?: RetrievalMethod[];

  query_analysis: QueryAnalysis | null;

  query_classification: QueryClassification | null;

  evidence: Evidence[];

  provenance?: Provenance[];

  graph_evidence?: GraphEvidence[];

  metadata: RetrievalMetadata;
}

/* -------------------------------------------------------------------------- */
/* Pipeline UI                                                                 */
/* -------------------------------------------------------------------------- */

export type RetrievalStage =
  | "idle"
  | "analyzing"
  | "classifying"
  | "planning"
  | "retrieving"
  | "semantic_retrieval"
  | "keyword_retrieval"
  | "graph_retrieval"
  | "fusing_results"
  | "reranking"
  | "collecting_evidence"
  | "resolving_provenance"
  | "completed"
  | "error";

export interface RetrievalPipelineStatus {
  stage: RetrievalStage;

  message?: string | null;

  progress?: number | null;

  completed?: boolean;

  error?: string | null;
}

/* -------------------------------------------------------------------------- */
/* Source citation                                                             */
/* -------------------------------------------------------------------------- */

export interface SourceCitation {
  paper_id: UUID | null;

  document_id: UUID | null;

  section_id?: UUID | null;

  chunk_id: UUID;

  page: number | null;

  section: string | null;

  source: string | null;

  graph_node_id?: UUID | null;

  graph_node_type?: GraphNodeType | null;

  graph_edge_id?: UUID | null;

  graph_edge_type?: string | null;
}

export function provenanceToCitation(
  provenance: Provenance,
): SourceCitation {
  return {
    paper_id: provenance.paper_id,

    document_id: provenance.document_id,

    chunk_id: provenance.chunk_id,

    page: provenance.page,

    section_id: provenance.section_id ?? null,

    section: provenance.section,

    source: provenance.source,

    graph_node_id: provenance.graph_node_id ?? null,

    graph_node_type: provenance.graph_node_type ?? null,

    graph_edge_id: provenance.graph_edge_id ?? null,

    graph_edge_type: provenance.graph_edge_type ?? null,
  };
}

/* -------------------------------------------------------------------------- */
/* Retrieval statistics                                                        */
/* -------------------------------------------------------------------------- */

export interface RetrievalStatistics {
  semantic_count: number;

  keyword_count: number;

  graph_count: number;

  fused_count: number;

  reranked_count: number;

  evidence_count: number;

  average_score?: number | null;

  graph_coverage?: number | null;

  metadata?: RetrievalMetadata;
}